uiowa/uiowa
# [Facilities] Project data

Resolves https://github.com/uiowa/uiowa/issues/6447. 
Resolves https://github.com/uiowa/uiowa/issues/7278. 
Resolves #7280 

# How to test
```
 ddev composer install && ddev blt ds --site=facilities.uiowa.edu && ddev drush @facilities.local facilities_core:projects_import && ddev drush @facilities.local uli /all-projects
```
1. Code review
2. Should take around 3:30 to import the projects. 
3. Run `ddev drush @facilities.local facilities_core:projects_import` again and should take around 2:40. 
4. Confirm dates are being imported correctly, compare https://buildui.facilities.uiowa.edu/buildui/ext/projectinfo?projnumber=0874301 to https://facilities.uiowa.ddev.site/project/john-pappajohn-business-building-upgrade-hvac-controls-system using https://www.epochconverter.com/. 
5. Confirm API call is bringing together all projects for import based on the list below. 
6. On `/all-projects` there should be 91 capital projects and 16 featured projects
7. Test a few buildings on `/all-projects` and compare against https://www.facilities.uiowa.edu/projects.  


1.   Grab all projects and data based on existing node ids from https://buildui.facilities.uiowa.edu/buildui/ext/projects?bldgnumber=0001 for example. 
2.   Grab all capital projects https://buildui.facilities.uiowa.edu/buildui/ext/capitalprojects
3.   Grab all featured projects https://buildui.facilities.uiowa.edu/buildui/ext/featuredprojects
4.   Finally, grab all additional project data from https://buildui.facilities.uiowa.edu/buildui/ext/projectinfo?projnumber=0897601

## Outstanding issues?
- ~Projects without an entity reference for a building are not importing the dates for some reason. `Parking Lots and Ramps - 2024 Surface Lot Maintenance` is an example: https://buildui.facilities.uiowa.edu/buildui/ext/projectinfo?projnumber=1062101.~
- https://facilities.uiowa.ddev.site/project/biomedical-research-support-facility-expand-facility imports even though https://buildui.facilities.uiowa.edu/buildui/ext/projectinfo?projnumber=0941301 is empty. 
- Do we need to find the total count of projects from the endpoint and compare that against what is imported?
- What determines the updated nodes when `ddev drush @facilities.local facilities_core:projects_import` is run again?  `Facilities projects content sync completed. 233 projects were created, 0 updated, 0 deleted, 0 skipped. That is neat.` is the first time message and ` 0 projects were created, 101 updated, 0 deleted, 132 skipped. That is neat.` the second time.

## Test themed page
- Project number 1062101 is the only one with a pre bid date greater than today
- Project number 0941301 estimated cost should output as "TBD" since the value is `0`.
- Add an image, webcam link, description, and what's new to an existing project. 
- Confirm two column layout has fields listed in the expected order https://www.facilities.uiowa.edu/project/0852601

## Test projects on building pages
- Confirm projects associated with a building appear on that building page.
- Example: https://facilities.uiowa.ddev.site/building/0204 should list several projects
- Check that the link to a given project works correctly.
- Confirm a building with no active projects does not list any. The Projects headline should also not display on these pages.
- Example: https://facilities.uiowa.ddev.site/building/0005
